### ISSUE TITLE
Update Dawn to use the Cart Ajax API instead of undocumented endpoints

### ISSUE THREAD

**Issue by bakura10** [Shopify/dawn#3207]
Hi,

For a few months, we have been experiencing **a ton** of issues and regressions, happening at the Cart Ajax API level, sometimes on super basic features like "removing a line item from cart" which fails. We are have found today such a regression on a merchant store. The Cart Ajax API has been extremely unreliable to use lately.

Those issues are extremely frustrating to handle on our end because despite being Shopify issues, merchants say that they reached Shopify, that Shopify support tried on Dawn, and because it works on Dawn, then it must be a theme issue (and, of course, merchants will always tend to believe Shopify rather than us).

One thing that exhibits the problem is that Dawn itself uses **undocumented APIs** that appears to follow different paths on Shopify backend (and that seems to work while the documented API does not). If Dawn theme would use the documented API, it would help a lot to demonstrate to merchants that the issue is indeed due to Shopify and not the theme. Shopify eating its own food would also ensure that such cart issues are spotted and fixed faster.

For instance, when removing a product from the cart will do a POST Ajax request to the endpoint `/cart/change`, with a JSON payload:

<img width="573" alt="image" src="https://github.com/Shopify/dawn/assets/1198915/75b9c108-02ee-468c-badf-df19a9c2b575">

<img width="703" alt="image" src="https://github.com/Shopify/dawn/assets/1198915/48390383-98f2-46af-a8c8-7cd816d1f132">

However, this API is undocumented. The only documented one is `/cart/change.js` (notice the .js at the end): https://shopify.dev/docs/api/ajax/reference/cart#post-locale-cart-change-js

Same thing when changing line item quantity, where Dawn uses undocumented `/cart/change` with JSON payload instead of `cart/change.js`:

<img width="559" alt="image" src="https://github.com/Shopify/dawn/assets/1198915/c43399ca-388f-4eab-97ba-1abf24aef119">

When adding to the cart, Dawn also uses an Ajax call to `/cart/add` instead of the documented `cart/add.js` endpoint (although, contrary to the line item removal, for this one Dawn uses a form data so it would be the same as submitting the form without Ajax ;but still, as soon as you're doing Ajax call Dawn should use the documented APIs for that work).

**Comment by resistorsoftware** [comment]
Did you ever ask yourself what the dotjs means at the end of the url? Are you sure there is a difference between a POST to /cart/change and /cart/change.js? 

As an example of basic web services since a long time, it is entirely possible to add an extension to an endpoint, and that just tells the server how to format the output. In other words, if you add .js to the endpoint, the server formats your response knowing you're interested in JS responses. If you were to tack on change.html the server would know you're interested in HTML as your response. 

So instead of thinking there is some deep-state crazy going on here, you might want to check. Does a POST to /cart/change with date XYZ produce the exact same response as one to /cart/change.js? If not, why not? Perhaps in one case your data payload is formatted correctly, perhaps not in the other? 

As responsible web developers we go with these things. As a user of Shopify cart JS since inception, I do not think you are pointing out anything "undocumented" that would be directly causing you problems. Maybe you can point out something, but your current "effort" is not complete. It can be frustrating calling endpoints and not seeing the results you expect, but often this is due to minor technicalities like forgetting something, or formatting something slightly wrong.

**Comment by bakura10** [comment]
Yes, unfortunately the two endpoints are not alias. I remember in the past that I did some experiments and they are not interchangeable. For instance for the add to cart API, the way errors are returned differ (the Ajax API the one with .js does block if you try to add more items to the cart than the inventory, while the non .js URL will not).

For the non .js API there are also some very odd quirks where you have to add a X-Requested-With header.

They are definitely differences between the two, that justify enough for dawn to use the documented endpoints! The simple fact that Shopify support uses Dawn as a reference to check if « it is a theme issue » and blame our theme if it works on Dawn, is an enough justification to align Dawn to only supports the documented URL.

Regarding the URL I don’t know in rails, but all the backend stacks I have worked (Zend, Expressive…) you specify the full name of the routes so if you define a route « cart/add » it won’t match « cart/add.js ». This might be different here but I don’t have access to their backend code to confirm :)

**Comment by resistorsoftware** [comment]
I agree you've discovered some facets of networking and making correct HTTP calls, but fundamentally, what I pointed out to you, is that indeed, the endpoint is the same. Same data in, same data out. What you are carefully documenting and discovering, is that there are certain conditions and criteria present when using these HTTP calls. 

I will leave you with your discoveries, but for the record, there is no underlying actual differences in these calls. The service provided is consistent. The black box of the **update** endpoint does not function or behave different. It does require slightly altered calling signature setup work from the developer, but that is not the same issue.

**Comment by bakura10** [comment]
Hello gain @ludoboludo ,

This issue is still valid, and since one week Shopify has introduced a major regression to the Ajax API. It is no longer possible to remove products from the cart when using the `/cart/change.js` Ajax endpoint when the merchant uses a market that contains both a country code and language code (for instance mysite.com/en-fr).

The problem is that the bug only appears in the `cart/change.js` (the documented endpoint), but not in the /cart/change Ajax endpoint used by Dawn (which is undocumented and that I've always been reluctant to use for this reason).

Dawn using an undocumented endpoint makes our life harder at support, as Shopify team always try on Dawn and concludes that, because it works on Dawn, it is a theme issue, while most cart issues are 100% on Shopify side.

If Dawn would use the documented endpoints, such issues would be spotted much faster.
